tests/lib/tools/tests.device-cgroup

```diff
@@ -0,0 +1,210 @@
+    deny.add_argument("major_minor", help="major:minor")
+    deny.set_defaults(func="deny")
+
+    opts = parser.parse_args()
+    if not hasattr(opts, "func"):
+        parser.print_help()
+        raise SystemExit(1)
+    return opts
+
+
+def is_cgroup_v2():
+    CGROUP2_SUPER_MAGIC = "0x63677270"
+    output = subprocess.check_output(
+        ["stat", "-f", "-c", "0x%t", "/sys/fs/cgroup"],
+    )
+    fstype = output.decode().strip()
+    logging.debug("/sys/fs/cgroup FS type %s", fstype)
+    return fstype == CGROUP2_SUPER_MAGIC
+
+
+def dump_v1(snap_app):
+    """Dump device entries from a v1 device cgroup. Entries are already formatted
+    in the correct way"""
+    group_name = "snap.{}".format(snap_app)
+    with open(
+        os.path.join("/sys/fs/cgroup/devices/", group_name, "devices.list")
+    ) as inf:
+        sys.stdout.write(inf.read())
+
+
+def device_change_v1(snap_app, allow, kind, major_minor):
+    group_name = "snap.{}".format(snap_app)
+    devices_control = "devices.allow"
+    if not allow:
+        devices_control = "devices.deny"
+    with open(
+        os.path.join("/sys/fs/cgroup/devices/", group_name, devices_control), "w+"
+    ) as outf:
+        outf.write("{} {} rwm".format(kind, major_minor))
+
+
+class DeviceCgroupV2Key(ctypes.Structure):
+    """A class representing v2 device cgroup key entry, which is 9 bytes long"""
+
+    # UINT32_MAX
+    DEVICE_MINOR_ANY = 4294967295
+
+    _pack_ = 1
+    _fields_ = [
+        # type is already a correct v1-like char 'b', 'c'
+        ("_typ", ctypes.c_char),
+        ("major", ctypes.c_uint32),
+        ("minor", ctypes.c_uint32),
+    ]
+
+    @property
+    def typ(self):
+        return self._typ.decode()
+
+
+def dump_v2(snap_app):
+    """Dump device entries for a v2 device cgroup. Entries are stored in a BPF map
+    created by snap-confine. We dump its contents using bpftool, and then
+    convert each entry to a format that looks like v1.
+    """
+    snap, app = snap_app.split(".")
+    map_pin_name = "snap_{}_{}".format(snap, app)
+    output = subprocess.check_output(
+        [
+            "bpftool",
+            "map",
+            "dump",
+            "pinned",
+            "/sys/fs/bpf/snap/{}".format(map_pin_name),
+            "-j",
+        ],
+    )
+    logging.debug("got output %s", output)
+    entries = json.loads(output)
+    for entry in entries:
+        raw_key = entry.get("key")
+        if not raw_key:
+            raise RuntimeError("unexpected object content in {}".format(output))
+        if len(raw_key) != ctypes.sizeof(DeviceCgroupV2Key):
+            raise RuntimeError(
+                "unexpected size of raw key {} (expected {})".format(
+                    len(raw_key), ctypes.sizeof(DeviceCgroupV2Key)
+                )
+            )
+        raw = bytearray([int(v, 16) for v in raw_key])
+        key = DeviceCgroupV2Key.from_buffer(raw)
+        # minor can be either the actual value, or a special value '*' which
```

Comment:
I would prepend "In cgroup v1," to this comment.

(note also the missing "s" in "represent" while you are at it :-) )